README.md
# Documentation

Published January 14, 2021


## Production Link

[Production Link](https://shakeriresearchgroup.vercel.app/)

## Summary

This project is a resource website which includes web applications and lab information for the Shakeri Research Group.

## Run this project locally

1. Make sure that Javascript is installed on your machine. You can install Node.js [here](https://nodejs.org/en/). 
2. Clone this repository ("git clone https://github.com/Shakeri-Lab/data-science-lab") using your terminal. If you are not using a unix OS (Mac or Linux) and don't already have a unix terminal emulator; [install git bash here](https://git-scm.com/downloads). 
3. cd into the project folder (data-science-lab) from your terminal
4. run "npm install"
5. run "npm start". A new browser tab will open displaying the local copy of the project.

## Push changes to the remote repository

1. cd into the project folder (data-science-lab) from your terminal
2. run "git pull" to get all te recent updates from the remote repository, otherwise conflicts can be introduced into the project.
3. run "npm test" to ensure that your changes work locally
4. run "git add *"
5. run "git commit -m "whatever_you_want_here"
6. run "git push"

## Edit the website Home Screen
### Site description (upper text)

Open src/Components/Home.md

### Alert (lower left text)

Open src/Components/Alert.md

### News (lower left text)

Open src/Components/News.md

## Edit the Blog Tab

Open src/Components/Blog.md

## Edit the Contact Tab

Open src/Components/Contact.md

## Add a personal bio

1. Inside src/Components/Personnel folder, create a file "FirstnameLastname.md". Add your bio content.
2. Inside src/Components/Personnel create a file "FirstnameLastname.js":

            import React, { useState, useEffect } from 'react';
            import ReactMarkdown from 'react-markdown'
            import MarkdownPath from "./FirstnameLastname.md"

            const FirstnameLastname = () => {
                var[text, setText] = useState(null)
                useEffect(() => {
                    fetch(MarkdownPath).then((response) => response.text()).then((text) => {
                        setText(text); 
                    })
                }, []);

                return(
                    <div className="App">
                        <div style={{margin: 20, textAlign: "left"}}>
                            <ReactMarkdown source={text} />
                        </div>
                    </div>
                )
            }

            export default FirstnameLastname; 

3. Open src/Components/Personnel.js. At line 36 inside the "BrowserView" and first "div" tags insert:

            <div style={{display: "flex", flexDirection: "column", backgroundColor: "whitesmoke", borderRadius: 10, margin: 20, width: 250, cursor: "pointer"}}
            onMouseClick={() => {
                window.location.href = "FirstnameLastname"
            }}
            >
                <img style={{borderRadius: 10, height: 250}} src="A_LINK_TO_YOUR_PROFILE_PHOTO" />
                <div style={{padding: 5}}>
                    <strong>Firstname Lastname</strong>
                    <p>WHAT YOU DO | INSTITUTION</p>
                </div>
            </div>

4. At line 63 inside the "MobileView" and first "div" tags insert:

             <div style={{borderRadius: 10, margin: 20}}
                onClick={() => {
                    window.location.href = "FirstnameLastname"
                }}
                >
                <img style={{borderRadius: 10, height: 250}} src="A_LINK_TO_YOUR_PROFILE_PHOTO"/>
                <div style={{padding: 5}}>
                    <strong>Firstname Lastname</strong>
                    <p>WHAT YOU DO | INSTITUTION</p>
                </div>
            </div>

5. Open src/App.js. Under the "//personnel" comment at line 25 insert:

        import FirstnameLastname from "./Components/Personnel/FirstnameLastname"; 

6. Under the "{/*Personnel*/}" comment at line 61 insert:

          <Route path="/FirstnameLastname" component={FirstnameLastname}>
            <FirstnameLastname />
          </Route>

## Add a new Project in the "Project" Tab

1. Inside src/Components/Projects folder, create a file "AProjectTitle.md". Add your project content.
2. Open src/Components/Projects.json. In the top of the JSON object add a new element in the following format:

            {"title": "Project_Title", "description": "A_description", "tags": "#project #tag", "url": "Atitle", "imageUrl": "anImageSourceURL"},
    
3. Open src/Components/Projects folder. Create a new file "AProjectTitle.js":

            import React, { useState, useEffect } from 'react';
            import ReactMarkdown from 'react-markdown'
            import MarkdownPath from "./AProjectTitle.md"

            const AProjectTitle = () => {
                var[text, setText] = useState(null)
                useEffect(() => {
                    fetch(MarkdownPath).then((response) => response.text()).then((text) => {
                        setText(text); 
                    })
                }, []);

                return(
                    <div className="App">
                        <div style={{margin: 20, textAlign: "left"}}>
                            <ReactMarkdown source={text} />
                        </div>
                    </div>
                )
            }

            export default AProjectTitle; 

4. Open src/App.js. Under the "//projects" comment at line 14 insert:

        import AProjectTitle from "./Components/Projects/AProjectTitle"; 

5. Under the "{/*Projects*/}" comment at line 43 insert:

        <Route path="/AProjectTitle" component={AProjectTitle}>
            <AProjectTitle />
        </Route>

## Add a new Publication in the "Publication" Tab

1. Inside src/Components/Publications folder, create a file "APublicationTitle.md". Add your publication content.
2. Open src/Components/Publications.json. In the top of the JSON object add a new element in the following format:

        {"title": "Publication_Title", "journal": "Journal Name", "year": "YYYY", "url": "Atitle", "description": "A_description", "imageUrl": "anImageSourceURL"},
    
3. Open src/Components/Publications folder and create a new file "APublicationTitle.js" and insert the following code:

            import React, { useState, useEffect } from 'react';
            import ReactMarkdown from 'react-markdown'
            import MarkdownPath from "./APublicationTitle.md"

            const APublicationTitle = () => {
                var[text, setText] = useState(null)
                useEffect(() => {
                    fetch(MarkdownPath).then((response) => response.text()).then((text) => {
                        setText(text); 
                    })
                }, []);

                return(
                    <div className="App">
                        <div style={{margin: 20, textAlign: "left"}}>
                            <ReactMarkdown source={text} />
                        </div>
                    </div>
                )
            }

            export default APublicationTitle; 

4. Open src/App.js. Under the "//publications" comment at line 19 insert:

        import APublicationTitle from "./Components/Publications/APublicationTitle"; 

5. Under the "{/*Publications*/}" comment at line 53 insert:

        <Route path="/APublicationTitle" component={APublicationTitle}>
            <APublicationTitle />
        </Route>

## Add a new Teaching/Lecture in the "Teaching/Lecture" Tab

1. Inside src/Components/Teaching folder, create a file "ATeachingTitle.md". Add your Teaching/Lecture content.
2. Open src/Components/Teaching.json. In the top of the JSON object add a new element in the following format:

        {"title": "Teaching/Lecture Title", "tags": "#teaching #tag", "date": "Jan 01, 2021", "url": "Atitle", "description": "A_description", "imageUrl": "anImageSourceURL"},
    
3. Open src/Components/Teaching folder. Create a new file "ATeachingTitle.js" and insert the following code:

        import React, { useState } from 'react';
        import ReactMd from 'react-md-file';

        const ATeachingTitle = () => {
            return(
                <div className="App">
                    <div className="content-container">
                        <ReactMd fileName="./ATeachingTitle.md" />
                    </div>
                </div>
            )
        }

        export default ATeachingTitle; 

4. Open src/App.js. Under the "//teaching" comment at line 22 insert:

        import Atitle from "./Components/Teachings/ATeachingTitle"; 

5. Under the "{/*Teaching/Lectures*/}" comment at line 57 insert:

        <Route path="/ATeachingTitle" component={ATeachingTitle}>
            <ATeachingTitle />
        </Route>

<hr></hr>

© 2020 Noah Beamon

public/News.md
<div style="text-align: left">
<h2>News</h2>

2021-01-06 The wastewater surveillance project has expanded outside of the university. Read <a href="https://dailyprogress.com/news/local/govt-and-politics/wastewater-surveillance-for-covid-expands-to-greater-charlottesville-area/article_532223c0-37d7-11eb-8268-cb15607ce35e.html#tracking-source=home-top-story%E2%80%8B" target="_blank">this article</a> from The Daily Progress to learn more.

2021-01-06 Check out <a href="https://www.nbc29.com/2020/12/13/uva-researchers-facilities-workers-test-wastewater-track-spread-covid-/" target="_blank">this article</a> by NBC29 regarding the use of wastewater by university researchers and facilities personnel to track the spread of COVID-19. 

2020-12-27 View our COVID-19 wastewater tracker web application <a href="/Wastewatertracker">here</a>. Want to learn more? Read the project documents <a href="/Wastewater">here</a>.

2020-09-14 Our project ``Risk Prediction for COVID-19'' with Vibrent Health is funded by NIH.

2020-08-01 Our project ``Epidemiologic Modeling, Public Health Surveillance and Sewershed'' is funded by Ivy Foundations.

</div>


src/Components/Home.js
import React, { useState, useEffect, useRef } from 'react';
import ReactMarkdown from 'react-markdown'
import MarkdownPath1 from "./Home.md"
import MarkdownPath2 from "./Alert.md"
import MarkdownPath3 from "./News.md"
import "./Home.css"; 
import { Tab, Tabs, TabList, TabPanel } from 'react-tabs';
import 'react-tabs/style/react-tabs.css';
import { TwitterTimelineEmbed } from 'react-twitter-embed';
import emailjs from 'emailjs-com';
import { AiFillGithub } from "react-icons/ai";
import { SiGooglescholar, SiOrcid} from "react-icons/si";
import RecentsJSON from "./Recents.json";
import { MdSubtitles } from "react-icons/md";
import { IoIosCalendar, IoMdSearch } from "react-icons/io";
import { BiHash } from "react-icons/bi"; 
import { FaTwitter } from "react-icons/fa";
import { Form, Button, FormGroup, FormControl, ControlLabel } from "react-bootstrap";

import {
    BrowserView,
    MobileView,
    isBrowser,
    isMobile
  } from "react-device-detect";

const Home = () => {
    // const[name, setName] = useState("");
    // const[email, setEmail] = useState("");
    // const[message, setMessage] = useState(""); 

    // const[button1, setButton1] = useState('2px solid rgb(0, 0, 0, 1)');
    // const[button2, setButton2] = useState('2px solid rgb(0, 0, 0, 1)');
    // const[button3, setButton3] = useState('2px solid rgb(0, 0, 0, 1)');

    // const[inputTitle, setInputTitle] = useState("");
    // const[inputDescription, setInputDescription] = useState("");
    // const[inputTags, setInputTags] = useState("");
    // const[inputTime, setInputTime] = useState("");

    // const[buttonColor, setButtonColor] = useState("rgba(46, 61, 130, 1)");

    var[text1, setText1] = useState(null)
    var[text2, setText2] = useState(null)
    var[text3, setText3] = useState(null)

    useEffect(() => {
        fetch(MarkdownPath1).then((response) => response.text()).then((text) => {
            setText1(text); 
        })
    }, []);

    useEffect(() => {
        fetch(MarkdownPath2).then((response) => response.text()).then((text) => {
            setText2(text); 
        })
    }, []);

    useEffect(() => {
        fetch(MarkdownPath3).then((response) => response.text()).then((text) => {
            setText3(text); 
        })
    }, []);



    const[twitterVisible, setTwitterVisible] = useState(false);
    const[alertVisible, setAlertVisible] = useState(true);

    return(
        <div className="App">
            {!isMobile && <BrowserView>
                <div style={{margin: 20, textAlign: "left"}}>
                    <ReactMarkdown source={text1} />
                </div>
                {alertVisible && <div style={{backgroundColor: "whitesmoke ", borderRadius: 10, margin: 20}}>
                    <ReactMarkdown source={text2} />
                    <p style={{color: "gray", cursor: "pointer"}}
                    onClick = {() =>{
                        setAlertVisible(false); 
                    }}><strong>X Close</strong></p>
                </div>}
                <div style={{display: "flex", flexDirection: "row", margin: 20}}>
                    <div style={{margin: 10, width: "50vw", height: window.innerHeight/1.5, overflow: "auto"}}>
                        <div style={{textAlign: "left"}}>
                            <ReactMarkdown source={text3} />
                        </div>
                    </div>
                    <div style={{margin: 10, width: "50vw"}}>
                    <TwitterTimelineEmbed
                        sourceType="profile"
                        screenName="HemanShakeri"
                        // options={{height: 500, width: 500}}
                        options={{height: 350}}
                    />
                    </div>
                </div>
            </BrowserView>}
            {isMobile && <MobileView style={{margin: 10}}>
                <div className="content-container">
                    <ReactMarkdown source={text1} />
                </div>
                {alertVisible && <div style={{backgroundColor: "whitesmoke ", borderRadius: 10, margin: 20}}>
                <ReactMarkdown source={text2} />
                    <p style={{color: "gray", cursor: "pointer"}}
                    onClick = {() =>{
                        setAlertVisible(false); 
                    }}><strong>X Close</strong></p>
                </div>}
                <div style={{margin: 10}}>
                    <FaTwitter style={{color: "#1197E8"}} size={30}
                     onClick={() => {
                        setTwitterVisible(!twitterVisible)
                     }}
                    />
                    {twitterVisible && <div>
                        <TwitterTimelineEmbed
                        sourceType="profile"
                        screenName="HemanShakeri"
                        options={{height: window.innerHeight/1.5}}
                    />
                    </div>}
                   
                </div>
                <div style={{margin: 10}}>
                    <ReactMarkdown source={text3} />
                </div>
            </MobileView>}

            {/* <Tabs style={{marginTop: 10}}>
                <TabList style={{display: "flex", alignItems: "left"}}>
                <Tab>General Info</Tab>
                <Tab>Activiy</Tab>
                <Tab>Announcements</Tab>
                <Tab>Feedback</Tab>
                </TabList>
                <TabPanel>
                    {alertVisible && <div style={{backgroundColor: "#b3e5fc", borderRadius: 10, margin: 20}}>
                        <p>Alert</p>
                        <p>We are currently accepting Graduate students. Please contact Heman Shakeri if you are interested. December 25, 2020 4:00 pm <p style={{color: "gray", cursor: "pointer"}}
                        onClick = {() =>{
                            setAlertVisible(false); 
                        }}><strong>X Close</strong></p></p>
                    </div>}
                    <div className="content-container">
                        <ReactMd fileName="./Home.md" />
                    </div>
                </TabPanel>
                <TabPanel>
                <p>Review activity or see the projects tab for more details.</p>
                <div style={{display: "flex", flexDirection: "row"}}>
                <div>
                    <div className = "content-results-container">
                    <strong style={{fontSize: 25, marginBottom: 10}}>Activity</strong>
                        <div className="scroll-container">
                            {RecentsJSON.filter((val)=>{
                                if (inputTitle == "" && inputTags == "" && inputDescription == "" && inputTime =="") {
                                    return val; 
                                }else if (val.title.toLowerCase().includes(inputTitle.toLowerCase()) && val.tags.toLowerCase().includes(inputTags.toLowerCase()) && val.date_time.toLowerCase().includes(inputTime.toLowerCase()) && val.description.toLowerCase().includes(inputDescription.toLowerCase())) {
                                    return val; 
                                }
                            }).map((val, key)=>{
                                return ( 
                                <div style={{borderRadius: "5px", cursor: "pointer"}} className = "list-element"
                                onClick = {() => {
                                    window.open(val.url);  
                                }}
                                >
                                    <div style={{display: "flex", flexDirection: "row"}}>
                                        <div>
                                            <strong>{val.title}</strong>
                                            <p>{val.description}</p>
                                            <p>tags: {val.tags}</p>
                                            <p style={{color: "gray"}}>{val.date_time}</p>
                                        </div>
                                        <div style={{marginLeft: 10}}>
                                            <img style={{width: 250, borderRadius: 5}} src={val.photo}/>
                                        </div>
                                    </div>
                                        
                                </div>)
                            })}
                        </div>
                </div>
                </div>
                <div style={{justifyContent: "center", alignItems: "center", marginTop: 0}}>
                <div className = "content-filter-container-a">
                    <strong style={{fontSize: 25}}>Filter Activity</strong>
                    <div style={{margin: 20}}>
                        <p>Title</p>
                        <div className="input-group">
                            <div className="input-group-prepend">
                                <span className="input-group-text" id="basic-addon">
                                <i className="fa fa-user prefix"><MdSubtitles/></i>
                                </span>
                            </div>
                            <input type="text" className="form-control" placeholder="title" value={inputTitle}
                            onChange={event => setInputTitle(event.target.value)}
                            />
                        </div>
                    </div>
                   <div style={{margin: 20}}>
                       <p>Keywords</p>
                    <div className="input-group">
                            <div className="input-group-prepend">
                                <span className="input-group-text" id="basic-addon">
                                <i className="fa fa-user prefix"><IoMdSearch/></i>
                                </span>
                            </div>
                            <input type="text" className="form-control" placeholder="description" value={inputDescription}
                            onChange={event => setInputDescription(event.target.value)}
                            />
                        </div>
                   </div>
                   <div style={{margin: 20}}>
                       <p>Tags</p>
                    <div className="input-group">
                            <div className="input-group-prepend">
                                <span className="input-group-text" id="basic-addon">
                                <i className="fa fa-user prefix"><BiHash/></i>
                                </span>
                            </div>
                            <input type="text" className="form-control" placeholder="tag" value={inputTags}
                            onChange={event => setInputTags(event.target.value)}
                            />
                        </div>
                   </div>
                   <div style={{margin: 20}}>
                       <p>Date/Time</p>
                    <div className="input-group">
                            <div className="input-group-prepend">
                                <span className="input-group-text" id="basic-addon">
                                <i className="fa fa-user prefix"><IoIosCalendar/></i>
                                </span>
                            </div>
                            <input type="text" className="form-control" placeholder="January 1, 2021 12:00 am" value={inputTime}
                            onChange={event => setInputTime(event.target.value)}
                            />
                        </div>
                   </div>
                   <div className="button" style={{backgroundColor: buttonColor, cursor: "pointer", borderRadius: 5, margin: 20}}
                            onMouseOver={() =>{
                                setButtonColor("rgba(46, 61, 130, 0.8)");
                            }}
                            onMouseLeave={() =>{
                                setButtonColor("rgba(46, 61, 130, 1)");
                            }}
                            onClick={() =>{
                                setInputTitle("");
                                setInputTags("");
                                setInputDescription("");
                                setInputTime(""); 
                                setButtonColor("rgba(46, 61, 130, 1)");
                            }}
                            >
                                <strong style={{color: "white"}}>Clear</strong>
                    </div>
                </div>
                <div>
                    <strong>Repositories and Databases</strong>
                </div>
                <div style={{display: "flex", flexDirection: "row"}}>
                    <div style={{border: button1, margin: 10, padding: 10, borderRadius: 5, cursor: "pointer", display: "flex", flexDirection: "row", justifyContent: "center", alignItems: "center"}}
                    onMouseOver={() => {
                        setButton1("2px solid rgb(0, 0, 0, 0.5)")
                    }}
                    onMouseLeave={() => {
                        setButton1("2px solid rgb(0, 0, 0, 1)")
                    }}
                    onClick={() => {
                        window.open('https://github.com/hshakeri'); 
                    }}
                    >
                        <AiFillGithub size={30} style={{marginRight: 10}}/>
                        <strong>Github</strong>
                    </div>
                    <div style={{border: button2, margin: 10, padding: 10, borderRadius: 5, cursor: "pointer", display: "flex", flexDirection: "row", justifyContent: "center", alignItems: "center"}}
                    onMouseOver={() => {
                        setButton2("2px solid rgb(0, 0, 0, 0.5)")
                    }}
                    onMouseLeave={() => {
                        setButton2("2px solid rgb(0, 0, 0, 1)")
                    }}
                    onClick={() => {
                        window.open('https://scholar.google.com/citations?user=zFIIhGMAAAAJ&hl=en'); 
                    }}
                    >
                        <SiGooglescholar size={30} style={{marginRight: 10}}/>
                        <strong>Google Scholar</strong>
                    </div>
                    <div style={{border: button3, margin: 10, padding: 10, borderRadius: 5, cursor: "pointer", display: "flex", flexDirection: "row", justifyContent: "center", alignItems: "center"}}
                    onMouseOver={() => {
                        setButton3("2px solid rgb(0, 0, 0, 0.5)")
                    }}
                    onMouseLeave={() => {
                        setButton3("2px solid rgb(0, 0, 0, 1)")
                    }}
                    onClick={() => {
                        window.open('https://orcid.org/0000-0002-9891-5748'); 
                    }}
                    >
                        <SiOrcid size={30} style={{marginRight: 10}}/>
                        <strong>ORCID</strong>
                    </div>
                </div>
                </div>
                </div>
                </TabPanel>
                <TabPanel>
                    <div style={{backgroundColor: "whitesmoke", marginLeft: 300, marginRight: 300, marginTop: 20, marginBottom: 20, paddingTop: 20, paddingBottom: 20, borderRadius: 5}}>
                        <TwitterTimelineEmbed
                        sourceType="profile"
                        screenName="HemanShakeri"
                        options={{height: 500, width: 500}}
                        />
                    </div>
                </TabPanel>
                <TabPanel>
                    <div>
                    <strong>Submit inquiries and suggestions, or report issues regarding our web applications here</strong>
                    <div style={{display: "flex", justifyContent: "center", margin: 20}}>
                    <Form style={{width: 500}} onSubmit={sendEmail}>
                        <Form.Group className="contact-form" >
                            <Form.Label>Email address</Form.Label>
                            <Form.Control type="email" placeholder="example@email.com"name="user_email" onChange={event => setEmail(event.target.value)}/>
                            <Form.Text className="text-muted">
                            We'll never share your email with anyone else.
                            </Form.Text>
                        </Form.Group>
                        <Form.Group className="contact-form">
                            <Form.Label>Name</Form.Label>
                            <Form.Control type="name" placeholder="name" name="user_name" onChange={event => setName(event.target.value)}/>
                            <Form.Text className="text-muted">
                            </Form.Text>
                        </Form.Group>
                        <Form.Group className="contact-form" >
                            <Form.Label>Leave a note</Form.Label>
                            <Form.Control as="textarea" type="name" placeholder="note" name="user_message" onChange={event => setMessage(event.target.value)}/>
                            <Form.Text className="text-muted">
                            </Form.Text>
                        </Form.Group>
                        <Form.File id="formcheck-api-regular">
                            <Form.File.Label>Upload a file</Form.File.Label>
                            <Form.File.Input name="user_file"/>
                        </Form.File>
                        <Button variant="primary" type="submit">
                            Submit
                        </Button>
                    </Form>
                    </div>
                    </div>
                </TabPanel>
            </Tabs> */}
        </div>
    )
}

export default Home; 